mixer/pkg/runtime/routing/builder.go

```diff
@@ -503,3 +524,133 @@ func (b *builder) getBuilderAndMapperDynamic(_ ast.AttributeDescriptorFinder,
 	}
 	return instBuilder, nil, nil
 }
+
+// chainedAttributeDescriptor extends an attribute descriptor with overrides
+type chainedAttributeDescriptor struct {
+	parent    ast.AttributeDescriptorFinder
+	overrides map[string]*descriptor.AttributeManifest_AttributeInfo
+}
+
+func (r *chainedAttributeDescriptor) GetAttribute(name string) *descriptor.AttributeManifest_AttributeInfo {
+	if info, exists := r.overrides[name]; exists {
+		return info
+	}
+	return r.parent.GetAttribute(name)
+}
+
+// buildRuleCompiler constructs an expression compiler over an extended attribute vocabulary
+// with template output attributes prefixed by the action names added to the global attribute manifests.
+func (b *builder) buildRuleCompiler(parent ast.AttributeDescriptorFinder, rule *config.Rule) *compiled.ExpressionBuilder {
+	// create attribute type descriptor from action templates
+	attributeDescriptor := make(map[string]*descriptor.AttributeManifest_AttributeInfo)
+
+	// TODO(kuat): add dynamic action support
+	for _, action := range rule.ActionsStatic {
+		if len(action.Instances) == 0 {
+			continue
+		}
+
+		// TODO(kuat): rationalize multi-instance check actions
+		template := action.Instances[0].Template
```

Comment:
I even missed where `template` (e.g.: authorization or metric templates) got a struct more robust than just string to earn `variety` and `attributemanifests`. After spending some time thinking, I didn't understand how an authorization instance could sometimes be considered CHECK and others CHECK WITH OUTPUT, because I don't remember finding the connexion between the instance template and the variety from any of the two PRs.
But if I could help with rational... maybe it would be just another `for` each instance.

Reply:
Each template has a fixed variety, which is either check or check with output. The only reason we have two, is that we cannot evolve one into the other, since gRPC signatures are not backwards compatible (which is caused by protos disallowing generic sub-protos). 

The runtime translate top-level `Check()` and `Report()` into `Dispath(variety)`. Here variety is slightly abused be either check or report variety. Check variety is then dispatched further to both check and check with output. One way to clean it up would be to decouple template variety from the dispatch variety. But I tried hard to refrain from any refactoring due to the size.